### ISSUE TITLE
NuGetSdkResolver adds significant overhead to evaluation and should not be used

### ISSUE THREAD

**Issue by davkean** [dotnet/msbuild#4025]
For evaluation time, The NuGetSdkResolver is extremely slow.

Based on investigation in CoreFX, I found:

- On restored state, it added 180-400ms overhead to a single project evaluation
- On unrestored state, it added 1000-6000ms overhead to a single project evaluation

I've looked [at the original design](https://github.com/Microsoft/msbuild/issues/2803), and it's built on a incorrect premise; that its performance impact will only affect the loading state of projects that use it. The image it shows where projects are being loaded in the background via Asynchronous Solution Load (ASL) was removed from the product in 15.5.

If a project opts into this resolver, it has the following effects:

- On solution load, blocks interaction with VS until it has completed downloading & resolving the SDK. 
- Increases the time it takes for the project-system to send design-time build results to Roslyn, resulting in delayed IntelliSense results
- Increases the time it takes for designers to open 
- Increases the time it takes for tests to be discovered
- Increases the time it takes for VS to react to changing the active configuration
- Increases the time and blocks interaction with VS when making changes to a project, such as adding/removing a file
- Increases design-time builds for every project that references it
- Increases the time for above when switching branches

When we designed SDK resolvers, it was explicitly called out that due to them being used during evaluation - that they _must be_ extremely fast and _must not_ hit the network. While this only hits the network on unrestored state, it still has a large negative impact on evaluation time when the package is already downloaded.

This is the entire reason that NuGet restore does not run during the build while inside Visual Studio.

Rogue resolvers can cause Visual Studio and other IDEs to be blamed for performance and UI delay issues, please remove or change the design of this resolver to play nicely with Visual Studio.

**Comment by livarcocc** [comment]
ACK on the negative performance impact. However, taking the feature out at this point is an impossibility.

Could we work together to figure out how to change the current design in a way to meet the perf requirements for VS?

**Comment by nguerrera** [comment]
Loooking at the code, https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs

It really doesn't do much in the restored state.

**Comment by davkean** [comment]
It was doing work, loading files off disk and parsing them if I remember correctly. I wish I could remember how I verified the overhead, I think I isolated in a single project and measured evaluation time.

**Comment by nguerrera** [comment]
![image](https://user-images.githubusercontent.com/75470/52251053-c3bf6900-28af-11e9-84e5-f0b21e9b0224.png)

Some quick thoughts:

It is spending the most time parsing global.json, which is sad because my global.json is 74 bytes.  We should look at switching to the new corefx json reader. There's a sources version for netfx IIRC. This file is trivial, we don't need a fancy deserializer. 

That said, it will avoid doing this more than once per build across projects:

https://github.com/NuGet/NuGet.Client/blob/b6cd1677ae2bd3b07f4cc23c2e8d408f784e8b05/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs#L89-L90

But I suspect VS evaluations aren't retaining this context object between projects. Fixing that if possible will help non-NuGet sdk evaluation considerably too.

Another major contributor is nuget loading its configuration files. I presume it needs to do this to answer the question of where a downloaded package is because values in the config can impact the answer to that. The parsing there can probably be improved as well.

For both of these we can possibly cache things in static state. and check config timestamps. I did some trickery like that in the non-nuget resolver.

**Comment by davkean** [comment]
This time matches almost identical to what I found in https://github.com/dotnet/sdk/issues/1483.

**Comment by nguerrera** [comment]
![image](https://user-images.githubusercontent.com/75470/52251752-9ffe2200-28b3-11e9-978f-0df83d44babd.png)

**Comment by davkean** [comment]
Nope, strike that, JSON.NET was way less than above.

**Comment by davidfowl** [comment]
We could just NGEN the one that comes with VS.

**Comment by nguerrera** [comment]
@davkean Different API used? This is a huge hammer to read global.json: https://github.com/NuGet/NuGet.Client/blob/b6cd1677ae2bd3b07f4cc23c2e8d408f784e8b05/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs#L66

**Comment by davkean** [comment]
Yep, just to grab a couple of strings from that file, hell I betcha regex would beat this hands down.

**Comment by nguerrera** [comment]
Yeah, there's a ReadAllText + IndexOf before it to avoid the deserialization, and that's not even showing up. 😆 

https://github.com/NuGet/NuGet.Client/blob/b6cd1677ae2bd3b07f4cc23c2e8d408f784e8b05/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs#L39-L46

**Comment by nguerrera** [comment]
I'll study some more tomorrow. This is interesting, but doesn't fully explain seeing 200ms - 400ms per project. When I'm seeing 200ms, of which 140ms is Jitting that would not have been happening N times for loading of full solution, right?

Mind you above you said: "I think I isolated in a single project and measured evaluation time". Possible your 200ms - 400ms included jitting?

**Comment by davkean** [comment]
When I was looking at traces for CoreFx, the resolver showing up as significant, enough to make me isolate this case and measure it by itself. Best way to measure this would be grab a real world use - such as the CoreFx solution I shared.

**Comment by nguerrera** [comment]
Yep. Will do that.

**Comment by nguerrera** [comment]
The remaining 60ms is still quite bad. Just wondering if the isolation process bloated it to 200ms. I will dig.

**Comment by davkean** [comment]
Bear in mind that the cost in that solution load will be ballooned by the number of evaluations (we evaluate every single project and configuration because the solution is incomplete) but it should be accurate to the cost of the total evaluation.

**Comment by jeffkl** [comment]
The parsed SDK versions from `global.json` are cached so it should only be read once.

https://github.com/NuGet/NuGet.Client/blob/b6cd1677ae2bd3b07f4cc23c2e8d408f784e8b05/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs#L81

MSBuild caches the resolved path for a particular SDK so that should only happen once. But for each unique SDK, the [NuGet settings are read](https://github.com/NuGet/NuGet.Client/blob/b6cd1677ae2bd3b07f4cc23c2e8d408f784e8b05/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs#L121).  That could be improved.

**Comment by jeffkl** [comment]
@nguerrera can you double check that if you have two or more projects referencing the same SDK-from-a-NuGet-package that there is no overhead except for one?  Also we still need to figure out if VS is using the EvaluationContext that enables caching of SDK resolution results.  If that hasn't been implemented yet, it would add a ton of overhead in VS.

**Comment by nguerrera** [comment]
I don't think VS is using the context, which is probably the biggest issue here.

**Comment by nguerrera** [comment]
Adding you to a thread Dave started on that.

**Comment by nguerrera** [comment]
Confirming that nuget setting reading is showing up. I mentioned that too with the trace. 

Even for happening once or once per sdk and even subtracting jit, it shouldn't need to take so long to read a few tiny files. Probably not what needs to be fixed for the large sln case though which I wil be studying. Above was just thoughts as they came to me looking at a trace for the first time.

**Comment by nguerrera** [comment]
The question for evaluation retaining context.State is what's the correct lifetime? It's not as simple as per build is, right? We have to text to global.json / project files changing at some point. For that reason I think we should attack both the cost on cache miss and hit. And separately ensure a good yet correct hit rate.

**Comment by Pilchie** [comment]
@cdmihai  and @lifengl have done some thinking about evaluation contexts in VS. We'd like to hook them up for the IO caching stuff too.

**Comment by rainersigwald** [comment]
There might be an easy win at "wrap a cache lifetime around solution load" where a ton of evaluations happen in a batch.

**Comment by cdmihai** [comment]
> There might be an easy win at "wrap a cache lifetime around solution load" where a ton of evaluations happen in a batch.

Whenever VS needs to evaluate a batch of projects would be good places to wrap in evaluation contexts, because if the disk changes then the result is undefined either way. Solution load is the most natural, but all APIs that receive a collection of projects to be evaluated / re-evaluated should do this.

There's also additional improvements that can be made to enhance the evaluation context:
- Seeding the contexts with partial views of the file system that VS has
- Setting them up in evaluations triggered by file based design time builds. The evaluations for these happen deep inside the engine, no way to currently inject an evaluation context in there.
- The ability to query the context for all the IO an evaluation did, so that VS can add extra file watchers and see when anything gets invalidated. But since this is not detours based tracking but rather tracking-as-convention, it gets more tricky

**Comment by davkean** [comment]
@rainersigwald Yeah I was thinking the same thing; how do I use said cache?

**Comment by cdmihai** [comment]
You use one of these new `Project` factory methods: https://github.com/Microsoft/msbuild/blob/master/src/Build/Definition/Project.cs#L526-L570
And pass the context in via `ProjectOptions`: https://github.com/Microsoft/msbuild/blob/master/src/Build/Definition/ProjectOptions.cs#L40

**Comment by jeromelaban** [comment]
On the same note, the `SdkResolverManifest` resolver:

https://github.com/Microsoft/msbuild/blob/b4f5792a8fadab87e5199ec54d355c3ba1397a6a/src/Build/BackEnd/Components/SdkResolution/SdkResolverManifest.cs#L27

is taking 236ms to initialize on my machine, which is more than the 153ms that the `NuGetSdkResolver` initialization is taking, mainly spent in the processing of the `DataContractSerializer`.

**Comment by nguerrera** [comment]
That can easily be converted to XmlReader, but does it happen more than once?

**Comment by nguerrera** [comment]
We should still change it, just curious if one time or many.

**Comment by jeromelaban** [comment]
The result seems cached, as it goes through `CachingSdkResolverService`, but it adds up to the rest. For instance, on the same profiling session, a single `new ProjectCollection()` takes 500ms (most of the time is spent searching for toolsets).

**Comment by davkean** [comment]
I was walking through some code today, and figured out that we do currently do use EvaluationContext during solution load if enabled via remote setting (let's us do an experiment). Figuring out if we actually turned this experiment on.

**Comment by davkean** [comment]
To update the thread, the team originally found no benefit to using evaluation context during load when we tested this - hence why it wasn't turned on (it could only break things). I suspect these resolvers might change this so we'll probably want to re-experiment on this, which I've filed to investigate: https://github.com/dotnet/project-system/issues/4649.

**Comment by davkean** [comment]
Here's a customer raising similar concerns on the symptoms I raised above: https://github.com/NuGet/Home/issues/8157.

**Comment by jeffkl** [comment]
Please review https://github.com/NuGet/NuGet.Client/pull/2879 to help me make this faster when reading global.json

**Comment by davkean** [comment]
Another internal partner reached out to me where they are experencing hangs and delays when reloading projects or adding a new file, and its caused the by same underlying cause.

**Comment by ViktorHofer** [comment]
Wanted to bump this. Just had a conversation with @davkean and he pointed to this as being one of the reasons why loading an sln with 500+ projects is terribly slow. (dotnet/runtime - libraries part)

@rainersigwald I think the assignee needs to be updated?

**Comment by CyrusNajmabadi** [comment]
This appears to be an issue causing hangs while VS is loading.  Can we get this off the UI codepath while loading?

**Comment by cdmihai** [comment]
> Can we get this off the UI codepath while loading?

True, there's two potential lines of work:
- make the nuget resolver faster by either making the resolver or nuget faster, or both
- don't do builds on the UI thread

**Comment by CyrusNajmabadi** [comment]
@cdmihai can this be prioritized.  This can literally block the UI thread for minutes *while VS is loading*.  Please see the linked dump.  This breaks VS entirely for the user.  It also violates a ton of best practices around using the UI thread.  

Can you explain what the nuget resolver does?  if it's anything related to IO, it sounds like the only option here is to get this off the UI thread.

**Comment by davkean** [comment]
@cdmihai This is not about doing builds on the UI thread, this is evaluation. We need to block on UI thread on evaluation because the user experience is really odd when you don't (you just added a file and it doesn't appear in the tree), or you just loaded a project and your files don't appear.

This isn't about making this feature faster, we were promised this extension point would never be used to hit the network which you can't avoid. We need to come up with an new extension point that designed with IDEs in mind. Given I've seen no progress on this, and these bugs are constantly pushed our way, I'm going to introduce new analysis that prompts a yellow bar when this feature slows down load of projects.

**Comment by jeffkl** [comment]
I'm trying to get some time to fix the perf issues in the case when the packages are already on disk:

- [ ] Use native library for loading `global.json` instead of a managed library (just like the .NET Core resolver does)
- [ ] NGen the resolver and load it by name instead of path in MSBuild to reduce JIT time
- [ ] Improve the perf of reading `NuGet.config` and cache the `NuGet.config` like we do with `global.json` so the file is only read once

The goal would be to make resolving an SDK from a NuGet package on par with the .NET Core SDK resolver.  Does that seem like the right direction?

To address the performance of when the resolver needs to pull a package, I want to add to the NuGet API so we can pull a single package faster.  Most SDK packages are very small so it should be quick.  https://github.com/NuGet/Home/issues/5919

Alternatively, I could add a new API to MSBuild that the project system could call to pre-resolve SDKs so they don't happen during evaluation.  That would be a bit more work but is entirely possible.

**Comment by davkean** [comment]
>Alternatively, I could add a new API to MSBuild that the project system could call to pre-resolve SDKs so they don't happen during evaluation. That would be a bit more work but is entirely possible.

Yes, this is the approach we should be taking - no matter how fast we make NuGet restore, it will never be fast enough.

**Comment by CyrusNajmabadi** [comment]
> I'm trying to get some time to fix the perf issues in the case when the packages are already on disk:

What happens when teh packages are not on disk?

**Comment by davkean** [comment]
It hits the network.

**Comment by jeromelaban** [comment]
@CyrusNajmabadi from what I remember, it can also freezes VS. I have to run `msbuild /r /t:Restore` to fix that.

**Comment by CyrusNajmabadi** [comment]
ok.  so network on hte UI thread basically has to be a complete no-no.  for one thing, i have *terrible* network access.  For anotehr, this blocks *loading* vs.  meaning i can't even do any work while this is happening because VS must be killed.

**Comment by dsplaisted** [comment]
I have a proposal to avoid the network access on the UI thread.  See the "NuGet SDK Resolver" section in this proposal: https://github.com/dotnet/designs/pull/104

Basically we are going to [update the MSBuild SDK resolver interface](https://github.com/microsoft/msbuild/issues/5239) so that a resolver can return no resolved SDK paths without failing.  A resolver will also be able to return items that should be added to the evaluation.

We can leverage this to fix the issue where the NuGet resolver downloads files on the UI thread.  In VS, the resolver should run in a mode that does not allow downloading SDKs.  Rather, if it can't find the SDK locally, it should return success and include a `MissingMSBuildSDK` item describing the SDK it would have downloaded.  The VS project system should check for those items on evaluation, and if there are any of them, it should spin off a non-UI thread task to download them.

**Comment by marcpopMSFT** [comment]
@dsplaisted with the optional workload work getting in, is it now possible for project system to update to handle this differently?

**Comment by dsplaisted** [comment]
This has been unblocked since the new SDK resolver support went in to MSBuild.  We would need to make changes to the [NuGet SDK Resolver](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver), and the project system would also need to make changes.  It would be nice to do that for 5.0.200 / VS 16.9.

**Comment by davkean** [comment]
This is another issue caused by it: https://developercommunity.visualstudio.com/t/VisualStudio-was-hang-when-open-the-wpf-/1429662. We have ~57 threads all blocked behind the NuGet resolver.

**Comment by davkean** [comment]
This is another issue caused by this: https://developercommunity.visualstudio.com/t/vs-performance-issue-ide-hangs-when-changing-file/1643242.

**Comment by aortiz-msft** [comment]
NuGet Epic to address perf issues in the NuGetSDKResolver: https://github.com/NuGet/Home/issues/11441.

Please feel free to reach out to @jeffkl directly.

**Comment by ladipro** [comment]
I'm looking at this issue, trying to break it down into units of work across the affected components. A couple of observations:

1. **Avoid downloading NuGet packages during evaluation**
I'm actually not sure if it should be built on top of the new SDK resolver support. If the SDK cannot be resolved, there is no point in continuing evaluation. Instead of defining items on the evaluated project to indicate which NuGet packages need to be restored, the resolver may as well just throw an exception with this info, which the project system would catch. Maybe I'm not understanding the flow correctly, though, will need to follow up.

The work needed here:
- From NuGet SDK resolver:
  - Detect that it's running in VS and it's not allowed to restore. The flag for VS is already there as `SdkResolverContext.IsRunningInVisualStudio` but presumably we'll want a way to opt-in/out to experiment so we may need an additional flag.
  - If the above holds true, indicate the unresolved package via an exception or via `ItemsToAdd` in the new SDK interface.
  - Overall, this is a fairly small work item.

- From MSBuild:
  - Depending on the exact mechanism of indicating the need to restore, MSBuild may need to aggregate the information from all resolver invocations. This is because `<Project Sdk="Package1/Version1;Package2/Version2">` is perfectly legal, whether intended to or not.
  - Very small amount of work.

- From the project system:
  - This would depend on the desired UX. We may want to have VS automatically restore the missing packages and then retry evaluating the affected project, all within the context of solution load. Or finish loading the solution leaving the affected projects 'unloaded' and then asynchronously pop a dialog? Or take the user directly to the VS NuGet Manager? Or something else. Very much an open question, feedback welcome.

2. **Avoid JITting**
There is indeed some non-trivial JITting cost when a process runs SDK resolution for the first time. For VS, the first project loaded after VS starts pays the price. On command-line it's generally every build (i.e. every MSBuild.exe invocation). On my machine I measured it at 45 ms for `Microsoft.Build.NuGetSdkResolver.dll` and 85 ms for `Microsoft.DotNet.MSBuildSdkResolver.dll`. The latter has nothing to do with NuGet SDK resolution but it's always invoked and because it's more expensive to JIT than the NuGet resolver, if we decide to address JITting it would make sense to prioritize fixing the base SDK resolver first.

I have played with prototyping a change to NGEN `Microsoft.Build.NuGetSdkResolver.dll` + its dependency `Microsoft.Deployment.DotNet.Releases` and it's not straightforward. MSBuild uses `Assembly.LoadFrom`, which immediately disqualifies the native image from being loaded. To be able to `Assembly.Load` by full name, we 1) Need to know the full name and 2) Need add a binding redirect to `devenv.exe.config` and `MSBuild.exe.config` so the CLR binder can find the file. For this we need to know the assembly version and in some cases this appears to be hitting layering & versioning unfortunacies, i.e. MSBuild generally doesn't know the specific .NET SDK version it's going to be part of. Not unsolvable but likely P2 in the grand scheme of things.

**Comment by dsplaisted** [comment]
> 1. I'm actually not sure if it should be built on top of the new SDK resolver support. If the SDK cannot be resolved, there is no point in continuing evaluation. Instead of defining items on the evaluated project to indicate which NuGet packages need to be restored, the resolver may as well just throw an exception with this info, which the project system would catch. Maybe I'm not understanding the flow correctly, though, will need to follow up.

The resolver needs to communicate this information to Visual Studio.  You could probably include it as data on an exception, but adding items seems cleaner to me.  Also if you do it via items, then I don't think you need to do the work in MSBuild to aggregate the information about which packages need to be downloaded, as there will just be multiple items with the information from the different resolver invocations.

**Comment by ladipro** [comment]
True, I think in general it's uglier to rely on structured data attached to an exception. On the other hand, if the resolver fails in this way, conceptually there's no value in evaluating the project. Other than the items added via `ItemsToAdd` it's pretty much useless. Perf wouldn't be an issue as this an error path. But what I'm kind of debating is whether a dummy `ProjectInstance` fits the CPS internal model or not. I.e. if it's good for CPS to always receive a valid object or if it's a potential source of confusion downstream. (ex: Would it be good or bad for such a `ProjectInstance` to be persisted to project system's evaluation cache?)

**Comment by BenjaminBrienen** [comment]
@ladipro are you still working on this?

**Comment by jeffkl** [comment]
I am hoping to work on the NuGet-based MSBuild project SDK resolver this year and am tracking work at https://github.com/NuGet/Home/issues/11441
